cpp/include/cudf/io/datasource.hpp

```diff
@@ -340,6 +340,10 @@ class arrow_io_source : public datasource {
    *
    * @param Apache Arrow Filesystem URI
    */
+
+  arrow_io_source()  = default;
```

Comment:
This class doesn't look like it has a sane default constructor, so it should be deleted, not defaulted. 
```suggestion
  arrow_io_source()  = delete;
```

Otherwise, something like this is ill formed:
```
arrow_io_source source{};
source.size(); // This will cause a segfault because the underlying `arrow_file` is uninitialized. 
```

Reply:
Does it make sense to allow a null (default-initialized) shared ptr there? Or do we consider a datasource not pointing to an underlying file to be in an invalid state?

Reply:
No idea what the expected behavior for a default constructed `arrow_io_source` should be.

Reply:
Okay - Thanks to @shwina 's suggestions, it seems that we no longer need the default ctor change in the PR anymore